Git error «ambiguous argument...unknown revision or path not in the working tree» when deploying

Have used the same setup before with success, but this time it refuses to do a deploy properly.

The error log:

```
--> Deploying to preprod environment
--> on host 10.3.2.16
  ○ deploying origin/preproduction
  ○ executing pre-deploy-local
  ○ hook pre-deploy
  ○ fetching updates
Fetching origin
  ○ resetting HEAD to origin/preproduction
fatal: ambiguous argument 'origin/preproduction': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

  git reset failed
```

The config for preprod (which fails):

```json
"preprod": {
            "user": "node",
            "host": "10.3.2.16",
            "ref": "origin/preproduction",
            "repo": "git@github.com:[something].git",
            "path": "/opt/www",
            "post-deploy": "npm install && pm2 reload pm2settings.json --env preprod"
        }
```

Note that I have the same config for the master and the production branch as well (to different servers), master works, preproduction and production fails with the message above.

Any tips?

@taktran I have the same configuration as you suggested but I am still getting the error when I try to do npm install a specific tagged version from a remote repo.

$ git config --list --show-origin
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=false
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        core.symlinks=false
file:.git/config        core.ignorecase=true
file:.git/config        remote.origin.url=ssh://git@repo.domain.co.uk:9999/proj/project.git
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.master.remote=origin
file:.git/config        branch.master.merge=refs/heads/master
file:.git/config        branch.dev.remote=origin
file:.git/config        branch.dev.merge=refs/heads/dev


When I do npm install i get he following error:
npm ERR! git rev-list -n1 v0.1.2: fatal: ambiguous argument 'v0.1.2': unknown revision or path not in the working tree.
npm ERR! git rev-list -n1 v0.1.2: Use '--' to separate paths from revisions, like this:
npm ERR! git rev-list -n1 v0.1.2: 'git <command> [<revision>...] -- [<file>...]'

Appreciate any help on this.

--- a/deploy
+++ b/deploy
@@ -207,13 +207,20 @@ setup() {
   local path=`config_get path`
   local repo=`config_get repo`
   local ref=`config_get ref`
+  local fetch=`config_get fetch`
   local branch=${ref#*/}
   hook_pre_setup || abort pre-setup hook failed
   run "mkdir -p $path/{shared/{logs,pids},source}"
   test $? -eq 0 || abort setup paths failed
   log running setup
   log cloning $repo
-  run "git clone --depth=5 --branch $branch $repo $path/source"
+  if test fetch != "fast"; then
+    log "full fetch"
+    run "git clone $repo $path/source"
+  else
+    log "fast fetch"
+    run "git clone --depth=5 --branch $branch $repo $path/source"
+  fi
   test $? -eq 0 || abort failed to clone
   run "ln -sfn $path/source $path/current"
   test $? -eq 0 || abort symlink failed
@@ -228,6 +235,7 @@ setup() {
 deploy() {
   local ref=$1
   local path=`config_get path`
+  local fetch=`config_get fetch`
 
   log "deploying ${ref}"
 
@@ -240,7 +248,15 @@ deploy() {
 
   # fetch source
   log fetching updates
-  run "cd $path/source && git fetch --depth=5 --all --tags"
+
+  if test fetch != "fast"; then
+    log "full fetch"
+    run "cd $path/source && git fetch --all --tags"
+  else 
+    log "fast fetch"
+    run "cd $path/source && git fetch --depth=5 --all --tags"
+  fi
+
   test $? -eq 0 || abort fetch failed
 
   # latest tag

--- a/deploy.js
+++ b/deploy.js
@@ -68,6 +68,9 @@ function deployForEnv(deploy_conf, env, args, cb) {
       ref: {
         type: 'string',
       },
+      fetch: {
+        type: 'string',
+      },
     },
     required: ["host", "repo", "path", "ref"],
   })) {

